### DISCUSSION TITLE
Should I Source Control public/mockServiceWorker.js

### DISCUSSION THREAD

**Discussion by jrock2004** [mswjs/msw#1015] [Q&A] [answered]
So I am wondering if I should be putting `public/mockServiceWorker.js` in source control or should I gitignore it?

**Comment by kettanaito** [comment, marked as answer]
Hey, @jrock2004. That's a great question.

Our stance on whether you should commit the worker script to your VCS has changed over time. In the past, the worker script was generated only by using the `npx msw init` command. That's why we used to recommend committing the script so that everybody working on your project would be sure they're using the same worker implementation.

Now, I recommend setting up the `msw.workerDirectory` property in your "package.json":

```json
{
  "msw": {
    "workerDirectory": "./public"
  }
}
```

With this property set, the worker script will automatically be generated whenever you install `msw`. This automatization allows you to treat the worker script as an artifact and _omit it from the VCS_.

**Reply by MattyBalaam** [reply]
Is this advice still current?

I have a monorepo using yarn 3.2.0 and I do not see a file created when running "yarn"

**Reply by kettanaito** [reply]
@MattyBalaam, as long as you specify `msw.workerDirectory` in your package.json (shown above), the worker script should be created at that directory whenever you run `yarn install`.

**Reply by MattyBalaam** [reply]
Unfortunately I'm not seeing this behaviour. I will do further investigation though. In our case the package.json is not the root one in our monorepo, could that be an issue?

**Reply by kettanaito** [reply]
Yes, that will absolutely be the issue. Our CLI only resolves the root-level package.json:

https://github.com/mswjs/msw/blob/7fb1a32bd1de2d75ff31d0c734c46d4b3bdabce2/cli/init.js#L54

So, if you want this to work in a monorepo, make sure you describe this property in the root-level package.json (alongside the `postinstall` script).

**Reply by MattyBalaam** [reply]
I could not get this to work, perhaps due to us using yarn 3 plug n play, so I have instead added a postinstall script to the child package to trigger this instead.

**Reply by kettanaito** [reply]
Yeah, different package managers, and even versions of the same manager, will respect `postinstall` differently. I've had some cases during testing where Yarn, I believe, would ignore `postinstall` script completely for some reason. Glad you've figured it out.

**Reply by tmskrtsz** [reply]
> Yes, that will absolutely be the issue. Our CLI only resolves the root-level package.json:
> 
> https://github.com/mswjs/msw/blob/7fb1a32bd1de2d75ff31d0c734c46d4b3bdabce2/cli/init.js#L54
> 
> So, if you want this to work in a monorepo, make sure you describe this property in the root-level package.json (alongside the `postinstall` script).

Is it possible to define multiple `"workerDirectory: {}"` paths? In my monorepo, there are multiple projects that would need the script.

**Reply by kettanaito** [reply]
Hi, @tmskrtsz. It's not currently possible but we have a feature request submitted (#1607) and this is one of the things I'd like to add after 2.0 lands later this month. Stay tuned.

**Reply by tmskrtsz** [reply]
Appreciate the speedy answer!

**Reply by alexbjorlig** [reply]
Hi @kettanaito

We are using Storybook msw addon. The `public` directory is not part of version control.
When I try to deploy to Vercel, everything seems to work but I get this error when I open the Storybook: `Did you forget to run "npx msw init <PUBLIC_DIR>"?`


I tried with this in the install script, `npx storybook build && npx msw init public` but it does not help. Do you have any idea what I can try to get it up and running?
![CleanShot 2024-02-01 at 14 38 13@2x](https://github.com/mswjs/msw/assets/9967422/66dce4ab-b5d8-4605-95b0-df27cc526bcd)
